### ISSUE TITLE
Disconnect use after free memory error in connectionLost_call

### ISSUE THREAD

**Issue by nhorman** [eclipse/paho.mqtt.c#1294]
**Describe the bug**
==2568== Memcheck, a memory error detector
==2568== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==2568== Using Valgrind-3.16.1 and LibVEX; rerun with -h for copyright info
==2568== Command: ./test
==2568== Parent PID: 2560
==2568==
==2568== Thread 7:
==2568== Invalid read of size 8
==2568== at 0x4CDA970: connectionLost_call (in /usr/lib/libpaho-mqtt3cs.so.1.3.9)
==2568== by 0x500AEA6: start_thread (pthread_create.c:477)
==2568== by 0x5121A2E: clone (clone.S:95)
==2568== Address 0x91e8238 is 40 bytes inside a block of size 176 free'd
==2568== at 0x48399AB: free (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2568== by 0x4CF8B0F: myfree (in /usr/lib/libpaho-mqtt3cs.so.1.3.9)
==2568== by 0x4CF3947: ListUnlink (in /usr/lib/libpaho-mqtt3cs.so.1.3.9)
==2568== by 0x4CF3A16: ListRemove (in /usr/lib/libpaho-mqtt3cs.so.1.3.9)
==2568== by 0x4CDA612: MQTTClient_destroy (in /usr/lib/libpaho-mqtt3cs.so.1.3.9)
==2568== by 0x142BC0: comms_run (comms_connect.c:436)
==2568== by 0x500AEA6: start_thread (pthread_create.c:477)
==2568== by 0x5121A2E: clone (clone.S:95)
==2568== Block was alloc'd at
==2568== at 0x483877F: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==2568== by 0x4CF8725: mymalloc (in /usr/lib/libpaho-mqtt3cs.so.1.3.9)
==2568== by 0x4CD9F58: MQTTClient_createWithOptions (in /usr/lib/libpaho-mqtt3cs.so.1.3.9)
==2568== by 0x1428FB: comms_run (comms_connect.c:404)
==2568== by 0x500AEA6: start_thread (pthread_create.c:477)
==2568== by 0x5121A2E: clone (clone.S:95)
==2568==

Problem was tracked down to the fact that MQTTClient_disconnect spawns a new thread to issue a connectionLost callback, which, if delayed, may mean the calling application can free/destroy the MQTTClient structure before the callback runs, causing the above use after free error above

**To Reproduce**

1. build a test program which connects an mqtt client to a broker, sets the connlost callback, disconnects via MQTTClient_disconnect and then immediately destroys the client
2. run under valgrind repeatedly, and eventually the the connlost callback thread will get scheduled behind the calling thread leading to a use after free condition in the mqtt libraries connectionLost_call function when m->cl is attempted to be dereferenced

**Expected behavior**
No use after free error

**Screenshots**
N/A

**Log files**
See description above for relevant valgrind log 

** Environment (please complete the following information):**
 - OS: Fedora 36

**Additional context**
Resolved by Pull request #1287

**Comment by icraggs** [comment]
You have to wait until all the disconnect processing is finished before destroying the client.

My intention was that if you called disconnect deliberately that connectionLost wasn't going to be called. I think I'll check that first.

**Comment by nhorman** [comment]
That....doesn't make any sense.  The disconnect processing is all internal to the library, external users have no visibility into when that is truly 'done'

Every disconnect call to MQTTClient_disconnect1 passes the value 1 in the third argument, so as long as the m->cl pointer is set, you'll clone a thread to make the connection lost callback. Maybe that wasn't the intent, but its certainly whats happening right now.

That said, It seems like a bad idea to require external users to wait for an internal library  process to complete before proceeding with a subsequent action.  I understand that external to the library a mutex could be inserted to block until a final connlost callback was received, but  the calling code may not be aware of what (if any) callback was registered for this particular client.   Even if it does, what outward example or guidance would the library give that was necessary?  My code called MQTTClient_disconnect, and received  a success status code.  If I were to write this code instead:
```
if (MQTTClient_disconnect5(client, ....) != SUCCESS) {
     /*Error handling and return */
}
if (MQTTClient_isConnected() == false) {
     MQTTClient_destroy(clent);
}
```
I would get the same valgrind error, because MQTTClient_isConnected() returns false immediately after the call to MQTTClient_disconnect5 completes.  If other library calls indicate that the disconnect is complete, why should I not trust that?  

Whats more, if I were to write this code:
```
if (MQTTClient_disconnect5(client...) == SUCCESS) {
     MQTTClient_setCallbacks(NULL, NULL, NULL);
}
```

I would very likely crash my program, because the disconnect call places my client struct in a state where I can modify the connection lost callback while there is still a potentially pending use that I may not be aware of as an external user.

Note also, the valgrind error I'm reporting isn't actually a crash, its just a use after free, which is really nasty to track down.  Theres no guarantee that this will immediately crash a system, because the address is valid, its just no longer allocated.  If the system in question happens to recycle this memory quickly enough, this would be a memory corruption bug.

If you didn't intend for the connection lost callback to be made, changing that would certainly fix the issue as well (though the code in question has been in place since 2012 it would seem, so I'm guessing someone relies on the callback occurring at this point).

It really seems like the better solution is to block internally on the connlost callback to ensure that any internal use of the client is complete by the time MQTTClient_disconnect1 returns.

An alternate solution might be to check for pending threads making callbacks in MQTTClient_destroy, but that would likely be a significantly more complex solution

**Comment by icraggs** [comment]
If deliberately calling disconnect doesn't cause connectionLost to be called, then you do know when it is done. The disconnect call is meant to be synchronous in the MQTTClient API, so either it should wait for the connectionLost callback to be completed, or not call the connectionLost callback. The latter is what I intended.

**Comment by nhorman** [comment]
But, neither of those things are happening, that's why I sent the PR.

MQTTClient_disconnect1 always spawns a thread to call the connectionLost
callback, anytime a connectionLost callback is registered.

As you note, I agree, that MQTTClient_disconect should be synchronous, but
it's not, as it never waits for the callback thread to complete, that's
what my PR fixes.  It creates a semaphore that the disconnect thread waits
on until the connectionLost callback thread posts it.

If you're preference is for MQTTCliet_disconnect to just never callback to
connectionLost, so be it, however:
1) the code to make the callback has been in MQTTClient_disconnect1 since
the initial commit to the repo in 2012
2) given (1), I would assume that there are applications that at this point
rely on the callback behavior when disconnect is called

So if it was your intent to never call connLost on disconnect, you are
welcome to make that fix instead of taking mine, however, I would strongly
suspect that you will break other users with that change.  It seems far
safer to just serialize the callback such that it's guaranteed complete by
the time disconnect returns (i.e. fix disconnect such that it is again
synchronous with the rest of the API) as my PR does

**Comment by icraggs** [comment]
The code is there to not call connectionLost on deliberate disconnect, and I set connectionLost in the paho_cs_pub.c sample to try it out, ending the sample with a crtl-C. ConnectionLost wasn't called in the times I tried it.

So I think that it was probably another reason for the disconnect prior to the application calling disconnect that caused the connectionLost to be called.

Regardless, it's a good idea for the disconnect1 function to synchronize on connectionLost completion though. Thanks for the PR.
